Decode general web exceptions

TheBeruriahIncident opened the pull request:
Previously, exceptions corresponding to specific HTTP error codes (e.g. BadRequestException) would be properly decoded, but more general exceptions that just hold a status value (e.g. ClientErrorException) would fail to decode.

We now handle:
- WebApplicationException
- ServerErrorException
- ClientErrorException

Fixes #24

markelliot commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -115,6 +118,25 @@ private static Exception constructException(
+                // WebApplicationException has a different order in its constructor arguments
+                if (exceptionClass.equals(WebApplicationException.class)) {
+                    return exceptionClass.getConstructor(Throwable.class, Integer.TYPE)
+                            .newInstance(wrappedException, status);
```
how about:

```
return new WebApplicationException(wrappedException, status);
```

markelliot commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -115,6 +118,25 @@ private static Exception constructException(
         // Note: If another constructor is added, then we should refactor the construction logic in order to avoid
         // nested try/catch
         try {
+            if (exceptionClass.equals(ClientErrorException.class)
```
this would also need to appear in the other `constructException` block for it to generally work, which makes me think the test coverage you're committing isn't sufficient.

markelliot replied:
or do these actually have a single-string arg constructor?

TheBeruriahIncident replied:
Ah, the other `constructException` was returning `RuntimeException`, will fix!

markelliot commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -115,6 +118,25 @@ private static Exception constructException(
+                // for example, "HTTP 501 Not Implemented"
+                int prefixLength = "HTTP ".length();
+                String statusAsString = message.substring(prefixLength, prefixLength + 3);
+                int status = Integer.parseInt(statusAsString);
```
I think maybe we should just put the regex in here, don't care about perf in the exception handler and would prefer something more readable

schlosna replied:
Would also avoid possible problems where statusAsString is non-numeric and won't parse.

uschi2000 commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -115,6 +118,25 @@ private static Exception constructException(
         // Note: If another constructor is added, then we should refactor the construction logic in order to avoid
```
I believe that this comment applies to this change. Introducing more special cases turns this into spaghetti.

uschi2000 replied:
Any opinion here?

TheBeruriahIncident replied:
That comment seems to be about the nested try-catch and a worry about nesting further. My change doesn't touch any of that, so I don't think it makes that situation worse. It may be possible to clean that block up, but I think that's a separate issue.

schlosna commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -115,6 +118,25 @@ private static Exception constructException(
+                // the message should be of the form "HTTP $STATUS $DESCRIPTION
+                // for example, "HTTP 501 Not Implemented"
+                int prefixLength = "HTTP ".length();
+                String statusAsString = message.substring(prefixLength, prefixLength + 3);
```
This may `throw IndexOutOfBoundsException` if the message doesn't contain "HTTP " or is length less than 5. Could also NPE if message is null.

schlosna commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -115,6 +118,25 @@ private static Exception constructException(
         try {
+            if (exceptionClass.equals(ClientErrorException.class)
+                    || exceptionClass.equals(ServerErrorException.class)
+                    || exceptionClass.equals(WebApplicationException.class)) {
```
Why not do `WebApplicationException.class.isAssignableFrom(exceptionClass)` to cover all of these cases or do you want to explicitly exclude `javax.ws.rs.RedirectionException`?

TheBeruriahIncident replied:
We want to exclude all of the Exception classes that were already handled. `BadRequestException`, `NotFoundException`, etc all have single-string constructors, but are children of WebApplicationException.

TheBeruriahIncident commented:
@markelliot , I made an oversight in that you can construct these exceptions in such a way that the message itself won't have the status, meaning that the status only lives as a field in the `Response`. As such, the `SerializableException` doesn't necessarily have enough info to actually reconstruct the exception. Is it reasonable to add a `status` field to `SerializableException`?

TheBeruriahIncident commented:
@markelliot Are you good with adding `status` to `SerializableException`?

uschi2000 commented:
@akroy why do you need it in the exception if it's already in the response?

TheBeruriahIncident commented:
@uschi2000 We don't currently put the status in the response. We've special cased a few exception types, but most everything just gets a generic 500 as the response status.

(I was planning on following this up with making an exception mapper that actually uses statuses) My (guessed) interpretation was that we were intentionally only using the `SerializableException` info for serialization, but if encoding the status in the response is fine, I'll work on our exception mappers first.

markelliot commented:
The response code is available from the literal http response, Adam, we shouldn't be additionally serializing it into the data structure.

TheBeruriahIncident commented:
Great. I'll close this out and add exception mapping that puts the status into the http response, and I'll reopen after that's fixed.

markelliot commented:
the status is already placed into the HTTP response.

TheBeruriahIncident commented:
Specific statuses are placed there for `FORBIDDEN` and `NOT_FOUND`, and we have handling for `IllegalArgumentException`, but from what I can tell/observe, most exceptions are handled by the `RuntimeExceptionMapper`, which just uses 500 no matter what.

markelliot commented:
- @pnepywoda

This doesn't quite jive with my recollection. I'm pretty sure if you return a WebAppException we pull out the error code.

TheBeruriahIncident commented:
I don't see any relevant mapper in: https://github.com/palantir/http-remoting/tree/develop/http-servers/src/main/java/com/palantir/remoting/http/server . I did write the (fairly small) PR to fix that though, I'll put it up in a sec

markelliot commented:
Ok, happy to look at the pr.

TheBeruriahIncident commented:
PR at https://github.com/palantir/http-remoting/pull/48

TheBeruriahIncident commented:
@markelliot @uschi2000 @schlosna Redid this now that we can assume the status is right

TheBeruriahIncident commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -112,7 +125,15 @@ private static Exception constructException(Class<? extends Exception> exception
+            return new ServerErrorException(status, wrappedException);
+        } else if (exceptionClass.equals(WebApplicationException.class)) {
+            return new WebApplicationException(wrappedException, status);
+        }
```
I'm not enthused about having this similar block twice, but since in one case we wrap the exception, more DRY approaches seem much less clean.

Similarly, I'm not a fan of having two different `if`s for Client and Server, but that lets us remove the reflection and take this out of the try-catch block.

uschi2000 replied:
you can `initCause()`; it's "clean enough" since it enforces (at runtime) that it's called at most once.

TheBeruriahIncident replied:
Just tried that: apparently all the WebApplicationExceptions call into a single constructor that takes cause, meaning that they're all setting cause to `null` if you don't set it. This prevents us from setting it with `initCause`, because of the "set once" limitation.

uschi2000 commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -101,7 +106,15 @@ public Exception decode(String methodKey, Response response) {
 
-    private static Exception constructException(Class<? extends Exception> exceptionClass, String message) {
+    private static Exception constructException(Class<? extends Exception> exceptionClass, String message, int status) {
+        if (exceptionClass.equals(ClientErrorException.class)) {
```
what about the message?

TheBeruriahIncident commented:
@uschi2000 Using message, improved tests

markelliot commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -101,7 +106,15 @@ public Exception decode(String methodKey, Response response) {
     }
 
-    private static Exception constructException(Class<? extends Exception> exceptionClass, String message) {
+    private static Exception constructException(Class<? extends Exception> exceptionClass, String message, int status) {
```
looking at http://grepcode.com/file/repo1.maven.org/maven2/javax.ws.rs/javax.ws.rs-api/2.0.1/javax/ws/rs/WebApplicationException.java#65 I think this method ought to simply invoke:

```
return constructException(exceptionClass, message, null, status);
```

...and that that would be a lot less duplicate code.

TheBeruriahIncident replied:
Collapsed the two methods into one

jlzhao27 commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -101,18 +106,16 @@ public Exception decode(String methodKey, Response response) {
-            return new RuntimeException(String.format(
-                    "Failed to construction exception of type %s, constructing RuntimeException instead: %s%n%s",
-                    exceptionClass.toString(), e1.toString(), message));
+    private static Exception constructException(
```
I don't think this change is correct. The problem I noticed is that the server side `JsonExceptionMapper` is capturing the web specific exceptions like 405 Not Allowed and throwing them as a 500. I think the fix needs to be in the exception mapper to throw the right status code (and we should also decode the exception).

TheBeruriahIncident replied:
Hey Jason! HTTP-remoting's handling of exceptions was actually doubly broken:
- most exceptions were thrown as 500s on server side (what you're mentioning)
- most exceptions were failing to decode on client side (what this commit is)

Both of these errors were related to the same failure to deal with WebApplication, ClientError, or ServerError exceptions, just in different places. I already fixed the problem you mention at https://github.com/palantir/http-remoting/pull/48 .

jlzhao27 replied:
oh doh, thanks adam! neither of these changes are released yet right?

uschi2000 replied:
nope. we can release 0.7.0 today. it's a major breaking change, both in
terms of code and config.

On Thu, Feb 18, 2016 at 2:54 PM Jason Zhao notifications@github.com wrote:

> In
> http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java
> https://github.com/palantir/http-remoting/pull/45#discussion_r53396621:
> 
> > @@ -101,18 +106,16 @@ public Exception decode(String methodKey, Response response) {
> >          }
> >      }
> > -    private static Exception constructException(Class<? extends Exception> exceptionClass, String message) {
> > -        try {
> > -            return exceptionClass.getConstructor(String.class).newInstance(message);
> > -        } catch (Exception e1) {
> > -            return new RuntimeException(String.format(
> > -                    "Failed to construction exception of type %s, constructing RuntimeException instead: %s%n%s",
> > -                    exceptionClass.toString(), e1.toString(), message));
> > -    private static Exception constructException(
> 
> oh doh, thanks adam! neither of these changes are released yet right?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/palantir/http-remoting/pull/45/files#r53396621.

TheBeruriahIncident replied:
Not yet! Server side fix is merged (so you could snapshot). On https://github.com/palantir/http-remoting/pull/35, @uschi2000 said that we'll likely release 0.7.0 once my fixes go in.

TheBeruriahIncident commented:
@markelliot @uschi2000 How does this look?

jlzhao27 commented:
ping on this also, GK is waiting on the 0.7.0 release :)

markelliot commented:
seems okay to me

TheBeruriahIncident commented:
Sweet. I'm going to squash and rebase

Commit:
Decode general web exceptions

Previously, exceptions corresponding to specific HTTP error codes (e.g. BadRequestException) would
  be properly decoded, but more general exceptions that just hold a status value (e.g. ClientErrorException)
  would fail to decode.

We now handle:
  - WebApplicationException
  - ServerErrorException
  - ClientErrorException

TheBeruriahIncident commented:
@markelliot Ready!

uschi2000 commented on http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java:
```diff
@@ -101,18 +106,16 @@ public Exception decode(String methodKey, Response response) {
-                    "Failed to construction exception of type %s, constructing RuntimeException instead: %s%n%s",
-                    exceptionClass.toString(), e1.toString(), message));
+    private static Exception constructException(
+            Class<? extends Exception> exceptionClass, String message, int status, Throwable wrappedException) {
```
@CheckForNull missing

uschi2000 replied:
or probably `Optional<Throwable>`

uschi2000 replied:
and please leave a comment saying that sub-class of RuntimeException can take a `null` cause parameter to indicate an unknown cause.

uschi2000 commented:
one last comment above, hopefully.
